README.md
# dynamic_color_theme

![](https://github.com/DFreds/dynamic_color_theme/blob/master/assets/demo.gif)

This package allows you to dynamically change your color theme and will automatically switch between light and dark mode based on the new color. Additionally, the color values can be persisted across restarts by saving both the color and dark mode to shared preferences.

## Installation

```yaml
dependencies:
    dynamic_color_theme: ^1.1.0
```

Import the theme as shown below.

```dart
import 'package:dynamic_color_theme/dynamic_color_theme.dart';
```

You can also utilize the built-in color picker dialog by importing as shown below.

```dart
import 'package:dynamic_color_theme/color_picker_dialog.dart';
```

## Usage

To use it, wrap your main widget like so:

```dart
const kFuchsia = const Color(0xFF880E4F);
const kWhite = Colors.white;
const kLightGrey = const Color(0xFFE8E8E8);
const kDarkGrey = const Color(0xFF303030);

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return DynamicColorTheme(
      data: (Color color, bool isDark) {
        return _buildTheme(color, isDark); // TODO define your own buildTheme method here
      },
      defaultColor: kFuchsia,
      defaultIsDark: false,
      themedWidgetBuilder: (BuildContext context, ThemeData theme) {
        return MaterialApp(
          home: MyHomePage(title: 'Dynamic Color Theme'),
          theme: theme,
          title: 'Flutter Demo',
        );
      },
    );
  }

  // Example buildTheme method
  ThemeData _buildTheme(Color accentColor, bool isDark) {
    final ThemeData base = isDark ? ThemeData.dark() : ThemeData.light();
    final Color primaryColor = isDark ? kDarkGrey : kWhite;

    return base.copyWith(
      accentColor: accentColor,
      accentTextTheme: _buildTextTheme(base.accentTextTheme, accentColor),
      cardColor: primaryColor,
      floatingActionButtonTheme: base.floatingActionButtonTheme.copyWith(
        backgroundColor: accentColor,
      ),
      iconTheme: base.iconTheme.copyWith(
        color: accentColor,
      ),
      primaryColor: primaryColor,
      primaryIconTheme: base.primaryIconTheme.copyWith(
        color: accentColor,
      ),
      primaryTextTheme: _buildTextTheme(base.primaryTextTheme, accentColor),
      scaffoldBackgroundColor: primaryColor,
      textSelectionTheme: _buildTextSelectionTheme(base.textSelectionTheme, accentColor, isDark),
      textTheme: _buildTextTheme(base.textTheme, accentColor),
    );
  }

  TextTheme _buildTextTheme(TextTheme base, Color color) {
    return base.copyWith(
      bodyText2: base.bodyText2.copyWith(
        fontSize: 16,
      ),
      bodyText1: base.bodyText1.copyWith(
        color: color,
        fontSize: 16,
        fontWeight: FontWeight.bold,
      ),
      button: base.button.copyWith(
        color: color,
      ),
      caption: base.caption.copyWith(
        color: color,
        fontSize: 14,
      ),
      headline5: base.headline5.copyWith(
        color: color,
        fontSize: 24,
      ),
      subtitle1: base.subtitle1.copyWith(
        color: color,
        fontSize: 18,
      ),
      headline6: base.headline6.copyWith(
        color: color,
        fontSize: 20,
        fontWeight: FontWeight.bold,
      ),
    );
  }

  TextSelectionThemeData _buildTextSelectionTheme(TextSelectionThemeData base, Color accentColor, bool isDark) {
    return base.copyWith(
      cursorColor: accentColor,
      selectionColor: isDark ? kDarkGrey : kLightGrey,
      selectionHandleColor: accentColor,
    );
  }
}
```

You can change the theme yourself by utilizing the provided functions as shown below.

```dart
void changeColor(Color newColor) {
  DynamicColorTheme.of(context).setColor(
    color: newColor,
    shouldSave: true, // saves it to shared preferences
  );
}

void changeIsDark(bool isDark) {
  DynamicColorTheme.of(context).setIsDark(
    isDark: isDark,
    shouldSave: true, // saves it to shared preferences
  );
}

void resetTheme() {
  DynamicColorTheme.of(context).resetToSharedPrefsValues();
}
```

## Dialog Widget

![](https://github.com/DFreds/dynamic_color_theme/blob/master/assets/dialog.png)

Alternatively, a color picker dialog widget is included that will do this for you. Use it like so:

```dart

// With the barrier dismissible

showDialog(
  barrierDismissible: false,
  builder: (BuildContext context) {
    return ColorPickerDialog(
      defaultColor: kFuchsia,
      defaultIsDark: false,
    );
  },
  context: context,
);

// Without the barrier dismissible

showDialog(
  builder: (BuildContext context) {
    return WillPopScope(
      child: ColorPickerDialog(
        defaultColor: kFuchsia,
        defaultIsDark: false,
      ),
      onWillPop: () async {
        // Handles resetting if user taps off dialog
        DynamicColorTheme.of(context).resetToSharedPrefsValues();
        return true;
      },
    );
  },
  context: context,
);
```

CHANGELOG.md
## [1.2.0] - June 21, 2021,

* Updated the Plugin to Null-Safety

## [1.1.2] - March 25, 2021.

* Update the shared preferences and color picker library
* Update the example to remove deprecated widgets and properties

## [1.1.1] - July 13, 2020.

* Update the shared preferences and color picker library

## [1.1.0] - February 17, 2020.

* Update the color picker dialog library
* Add configuration for showing the label in the built-in color dialog

## [1.0.0] - February 13, 2020.

* DynamicColorTheme initial release
* ColorPickerDialog initial release


lib/color_picker_dialog.dart
import 'package:dynamic_color_theme/dynamic_color_theme.dart';
import 'package:flutter/material.dart';
import 'package:flutter_colorpicker/flutter_colorpicker.dart';

class ColorPickerDialog extends StatefulWidget {
  ColorPickerDialog({
    @required this.defaultColor,
    @required this.defaultIsDark,
    this.title = 'Choose a Color',
    this.cancelButtonText = 'CANCEL',
    this.confirmButtonText = 'CONFIRM',
    this.shouldAutoDetermineDarkMode = true,
    this.shouldShowLabel = false,
  });

  /// The default color when reset is tapped in the dialog.
  final Color defaultColor;

  /// The default dark mode when reset is tapped in the dialog.
  final bool defaultIsDark;

  /// The title of the dialog.
  ///
  /// By default, it is set to 'Choose a Color'.
  final String title;

  /// The text to display for the cancel button.
  ///
  /// By default, it is set to 'CANCEL'.
  final String cancelButtonText;

  /// The text to display for the confirm button.
  ///
  /// By default, it is set to 'CONFIRM'.
  final String confirmButtonText;

  /// Determines if the dark mode should be set based on the chosen color.
  ///
  /// By default, it is set to true.
  final bool shouldAutoDetermineDarkMode;

  /// Determines if the picker should show the HEX/RGB/HSV/HSL values.
  ///
  /// By default, it is set to false.
  final bool shouldShowLabel;

  @override
  _ColorPickerDialogState createState() => _ColorPickerDialogState();
}

class _ColorPickerDialogState extends State<ColorPickerDialog> {
  bool isDark;
  Color pickedColor;

  @override
  void initState() {
    super.initState();

    isDark = DynamicColorTheme.of(context).isDark;
    pickedColor = DynamicColorTheme.of(context).color;
  }

  @override
  Widget build(BuildContext context) {
    return AlertDialog(
      actions: <Widget>[
        InkWell(
          onTap: () {
            DynamicColorTheme.of(context).resetToSharedPrefsValues();
            Navigator.of(context).pop();
          },
          child: Padding(
            padding: const EdgeInsets.all(16),
            child: Text(widget.cancelButtonText),
          ),
        ),
        InkWell(
          onTap: () {
            DynamicColorTheme.of(context).setColor(
              color: pickedColor,
              shouldSave: true,
            );
            DynamicColorTheme.of(context).setIsDark(
              isDark: isDark,
              shouldSave: true,
            );

            Navigator.of(context).pop();
          },
          child: Padding(
            padding: const EdgeInsets.all(16),
            child: Text(widget.confirmButtonText),
          ),
        ),
      ],
      content: SingleChildScrollView(
        child: ColorPicker(
          displayThumbColor: true,
          enableAlpha: false,
          showLabel: widget.shouldShowLabel,
          onColorChanged: (Color color) {
            setState(() {
              pickedColor = color;

              if (widget.shouldAutoDetermineDarkMode) {
                isDark = !useWhiteForeground(color);
              }
            });

            DynamicColorTheme.of(context).setColor(
              color: color,
              shouldSave: false,
            );

            DynamicColorTheme.of(context).setIsDark(
              isDark: isDark,
              shouldSave: false,
            );
          },
          pickerAreaHeightPercent: 0.7,
          pickerColor: pickedColor,
        ),
      ),
      title: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: <Widget>[
          Text(widget.title),
          IconButton(
            icon: Icon(Icons.settings_backup_restore),
            onPressed: _reset,
            tooltip: "Reset to Default",
          ),
        ],
      ),
    );
  }

  void _reset() {
    setState(() {
      pickedColor = widget.defaultColor;
      isDark = widget.defaultIsDark;
    });

    DynamicColorTheme.of(context).setColor(
      color: widget.defaultColor,
      shouldSave: false,
    );

    DynamicColorTheme.of(context).setIsDark(
      isDark: isDark,
      shouldSave: false,
    );
  }
}


lib/dynamic_color_theme.dart
library dynamic_color_theme;

import 'dart:async';

import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';

typedef ThemedWidgetBuilder = Widget Function(
  BuildContext context,
  ThemeData data,
);

typedef ThemeDataBuilder = ThemeData Function(Color color, bool isDark);

class DynamicColorTheme extends StatefulWidget {
  const DynamicColorTheme({
    Key key,
    @required this.data,
    @required this.themedWidgetBuilder,
    @required this.defaultColor,
    @required this.defaultIsDark,
  }) : super(key: key);

  /// Function that provides the theme and expects a widget to be returned.
  ///
  /// Typically, this would return your MaterialApp widget.
  final ThemedWidgetBuilder themedWidgetBuilder;

  /// Function that provides the chosen color and dark mode selection and
  /// expects an instance of ThemeData to be returned.
  ///
  /// This data will be provided to the themedWidgetBuilder.
  final ThemeDataBuilder data;

  /// The default color to use if nothing has been saved in shared preferences
  /// yet.
  final Color defaultColor;

  /// The default dark mode selection if nothing has been saved in shared
  /// preferences yet.
  final bool defaultIsDark;

  @override
  DynamicColorThemeState createState() => DynamicColorThemeState();

  static DynamicColorThemeState of(BuildContext context) {
    return context.findAncestorStateOfType<DynamicColorThemeState>();
  }
}

class DynamicColorThemeState extends State<DynamicColorTheme> {
  ThemeData _data;

  Color _color;
  bool _isDark;

  static const String _colorSharedPreferencesKey = 'color';
  static const String _isDarkSharedPreferencesKey = 'isDarkMode';

  ThemeData get data => _data;

  Color get color => _color;

  bool get isDark => _isDark;

  @override
  void initState() {
    super.initState();
    _color = widget.defaultColor;
    _isDark = widget.defaultIsDark;

    _data = widget.data(_color, _isDark);

    Future<Color> colorFuture = loadColor();
    Future<bool> isDarkFuture = loadIsDark();

    Future.wait([colorFuture, isDarkFuture]).then((List results) {
      _color = results[0];
      _isDark = results[1];

      _data = widget.data(_color, _isDark);
      if (mounted) {
        setState(() {});
      }
    });
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    _data = widget.data(_color, _isDark);
  }

  @override
  void didUpdateWidget(DynamicColorTheme oldWidget) {
    super.didUpdateWidget(oldWidget);
    _data = widget.data(_color, _isDark);
  }

  /// Sets the provided color. If `shouldSave` is true, it will save it in
  /// shared preferences.
  ///
  /// This triggers a rebuild of the widgets utilizing the `DynamicColorTheme`.
  Future<void> setColor({
    @required Color color,
    @required bool shouldSave,
  }) async {
    setState(() {
      _data = widget.data(color, _isDark);
      _color = color;
    });

    if (shouldSave) {
      final SharedPreferences prefs = await SharedPreferences.getInstance();
      await prefs.setInt(_colorSharedPreferencesKey, color.value);
    }
  }

  /// Sets the dark mode. If `shouldSave` is true, it will save it in shared
  /// preferences.
  ///
  /// This triggers a rebuild of the widgets utilizing the `DynamicColorTheme`.
  Future<void> setIsDark({
    @required bool isDark,
    @required bool shouldSave,
  }) async {
    setState(() {
      _data = widget.data(_color, isDark);
      _isDark = isDark;
    });

    if (shouldSave) {
      final SharedPreferences prefs = await SharedPreferences.getInstance();
      await prefs.setBool(_isDarkSharedPreferencesKey, isDark);
    }
  }

  /// Resets the dark mode flag and color to the ones that are saved in shared
  /// preferences.
  ///
  /// This triggers a rebuild of the widgets utilizing the `DynamicColorTheme`.
  void resetToSharedPrefsValues() {
    Future<Color> colorFuture = loadColor();
    Future<bool> isDarkFuture = loadIsDark();

    Future.wait([colorFuture, isDarkFuture]).then((List results) {
      _color = results[0];
      _isDark = results[1];

      setState(() {
        _data = widget.data(_color, _isDark);
      });
    });
  }

  /// Loads the saved color from shared preferences.
  Future<Color> loadColor() async {
    final SharedPreferences prefs = await SharedPreferences.getInstance();
    int colorValue =
        prefs.getInt(_colorSharedPreferencesKey) ?? widget.defaultColor.value;
    return Color(colorValue);
  }

  /// Loads the saved dark mode flag from shared preferences.
  Future<bool> loadIsDark() async {
    final SharedPreferences prefs = await SharedPreferences.getInstance();
    return prefs.getBool(_isDarkSharedPreferencesKey) ?? widget.defaultIsDark;
  }

  @override
  Widget build(BuildContext context) {
    return widget.themedWidgetBuilder(context, _data);
  }
}


example/lib/main.dart
import 'package:dynamic_color_theme/color_picker_dialog.dart';
import 'package:dynamic_color_theme/dynamic_color_theme.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

const kFuchsia = const Color(0xFF880E4F);
const kWhite = Colors.white;
const kLightGrey = const Color(0xFFE8E8E8);
const kDarkGrey = const Color(0xFF303030);

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return DynamicColorTheme(
      data: (Color color, bool isDark) {
        return _buildTheme(color, isDark);
      },
      defaultColor: kFuchsia,
      defaultIsDark: false,
      themedWidgetBuilder: (BuildContext context, ThemeData theme) {
        return MaterialApp(
          home: MyHomePage(title: 'Dynamic Color Theme'),
          theme: theme,
          title: 'Flutter Demo',
        );
      },
    );
  }

  ThemeData _buildTheme(Color accentColor, bool isDark) {
    final ThemeData base = isDark ? ThemeData.dark() : ThemeData.light();
    final Color primaryColor = isDark ? kDarkGrey : kWhite;

    return base.copyWith(
      accentColor: accentColor,
      accentTextTheme: _buildTextTheme(base.accentTextTheme, accentColor),
      cardColor: primaryColor,
      floatingActionButtonTheme: base.floatingActionButtonTheme.copyWith(
        backgroundColor: accentColor,
      ),
      iconTheme: base.iconTheme.copyWith(
        color: accentColor,
      ),
      primaryColor: primaryColor,
      primaryIconTheme: base.primaryIconTheme.copyWith(
        color: accentColor,
      ),
      primaryTextTheme: _buildTextTheme(base.primaryTextTheme, accentColor),
      scaffoldBackgroundColor: primaryColor,
      textSelectionTheme: _buildTextSelectionTheme(base.textSelectionTheme, accentColor, isDark),
      textTheme: _buildTextTheme(base.textTheme, accentColor),
    );
  }

  TextTheme _buildTextTheme(TextTheme base, Color color) {
    return base.copyWith(
      bodyText2: base.bodyText2.copyWith(
        fontSize: 16,
      ),
      bodyText1: base.bodyText1.copyWith(
        color: color,
        fontSize: 16,
        fontWeight: FontWeight.bold,
      ),
      button: base.button.copyWith(
        color: color,
      ),
      caption: base.caption.copyWith(
        color: color,
        fontSize: 14,
      ),
      headline5: base.headline5.copyWith(
        color: color,
        fontSize: 24,
      ),
      subtitle1: base.subtitle1.copyWith(
        color: color,
        fontSize: 18,
      ),
      headline6: base.headline6.copyWith(
        color: color,
        fontSize: 20,
        fontWeight: FontWeight.bold,
      ),
    );
  }

  TextSelectionThemeData _buildTextSelectionTheme(TextSelectionThemeData base, Color accentColor, bool isDark) {
    return base.copyWith(
      cursorColor: accentColor,
      selectionColor: isDark ? kDarkGrey : kLightGrey,
      selectionHandleColor: accentColor,
    );
  }
}

class MyHomePage extends StatelessWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  Widget build(BuildContext context) {
    Color color = DynamicColorTheme.of(context).color;
    bool isDark = DynamicColorTheme.of(context).isDark;

    return Scaffold(
      appBar: AppBar(
        title: Text(title),
      ),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Center(
            child: Text('Color is $color and isDark is $isDark'),
          ),
          TextButton(
            child: Text(
              'Flip dark mode',
              style: Theme.of(context).textTheme.button,
            ),
            onPressed: () {
              DynamicColorTheme.of(context).setIsDark(
                isDark: !isDark,
                shouldSave: true,
              );
            },
          ),
          TextButton(
            child: Text(
              'Set color to Fuschia!',
              style: Theme.of(context).textTheme.button,
            ),
            onPressed: () {
              DynamicColorTheme.of(context).setColor(
                color: kFuchsia,
                shouldSave: true,
              );
            },
          ),
        ],
      ),
      floatingActionButton: FloatingActionButton.extended(
        backgroundColor: color,
        icon: Icon(Icons.color_lens),
        label: Text('Color Picker'),
        onPressed: () {
          showDialog(
            builder: (BuildContext context) {
              return WillPopScope(
                child: ColorPickerDialog(
                  defaultColor: kFuchsia,
                  defaultIsDark: false,
                  title: 'Choose your Destiny',
                  cancelButtonText: 'NEVERMIND',
                  confirmButtonText: 'SOUNDS GOOD',
                  shouldAutoDetermineDarkMode: true,
                  shouldShowLabel: true,
                ),
                onWillPop: () async {
                  DynamicColorTheme.of(context).resetToSharedPrefsValues();
                  return true;
                },
              );
            },
            context: context,
          );
        },
      ),
    );
  }
}
